pkg/services/ngalert/schedule/schedule.go

```diff
@@ -424,84 +422,110 @@ func (sch *schedule) ruleEvaluationLoop(ctx context.Context) error {
 }
 
 func (sch *schedule) ruleRoutine(grafanaCtx context.Context, key models.AlertRuleKey, evalCh <-chan *evalContext, stopCh <-chan struct{}) error {
-	sch.log.Debug("alert rule routine started", "key", key)
+	logger := sch.log.New("uid", key.UID, "org", key.OrgID)
+	logger.Debug("alert rule routine started")
 
-	evalRunning := false
-	var attempt int64
-	var alertRule *models.AlertRule
-	for {
-		select {
-		case ctx := <-evalCh:
-			if evalRunning {
-				continue
-			}
+	tenant := fmt.Sprint(key.OrgID)
+	evalTotalMetric := sch.metrics.EvalTotal.WithLabelValues(tenant)
+	evalDurationMetric := sch.metrics.EvalDuration.WithLabelValues(tenant)
+	evalTotalFailuresMetric := sch.metrics.EvalFailures.WithLabelValues(tenant)
 
-			evaluate := func(attempt int64) error {
-				start := timeNow()
+	updateRule := func() (*models.AlertRule, error) {
+		q := models.GetAlertRuleByUIDQuery{OrgID: key.OrgID, UID: key.UID}
+		err := sch.ruleStore.GetAlertRuleByUID(&q)
+		if err != nil {
+			logger.Error("failed to fetch alert rule", "error", err)
+			return nil, err
+		}
+		return q.Result, nil
+	}
 
-				// fetch latest alert rule version
-				if alertRule == nil || alertRule.Version < ctx.version {
-					q := models.GetAlertRuleByUIDQuery{OrgID: key.OrgID, UID: key.UID}
-					err := sch.ruleStore.GetAlertRuleByUID(&q)
-					if err != nil {
-						sch.log.Error("failed to fetch alert rule", "key", key)
-						return err
-					}
-					alertRule = q.Result
-					sch.log.Debug("new alert rule version fetched", "title", alertRule.Title, "key", key, "version", alertRule.Version)
-				}
+	evaluate := func(alertRule *models.AlertRule, attempt int64, ctx *evalContext) error {
+		logger := logger.New("version", alertRule.Version, "attempt", attempt, "now", ctx.now)
+		start := sch.clock.Now()
 
-				condition := models.Condition{
-					Condition: alertRule.Condition,
-					OrgID:     alertRule.OrgID,
-					Data:      alertRule.Data,
-				}
-				results, err := sch.evaluator.ConditionEval(&condition, ctx.now, sch.dataService)
-				var (
-					end    = timeNow()
-					tenant = fmt.Sprint(alertRule.OrgID)
-					dur    = end.Sub(start).Seconds()
-				)
-
-				sch.metrics.EvalTotal.WithLabelValues(tenant).Inc()
-				sch.metrics.EvalDuration.WithLabelValues(tenant).Observe(dur)
-				if err != nil {
-					sch.metrics.EvalFailures.WithLabelValues(tenant).Inc()
-					// consider saving alert instance on error
-					sch.log.Error("failed to evaluate alert rule", "title", alertRule.Title,
-						"key", key, "attempt", attempt, "now", ctx.now, "duration", end.Sub(start), "error", err)
-					return err
-				}
+		condition := models.Condition{
+			Condition: alertRule.Condition,
+			OrgID:     alertRule.OrgID,
+			Data:      alertRule.Data,
+		}
+		results, err := sch.evaluator.ConditionEval(&condition, ctx.now, sch.dataService)
+		dur := sch.clock.Now().Sub(start)
+		evalTotalMetric.Inc()
+		evalDurationMetric.Observe(dur.Seconds())
```

Comment:
Can we move this closer to the end of evaluate so that we're recording the time of the entire rule evaluation routine (including processing states and sending to AM)? Otherwise we're only capturing query time.

Reply:
Agree. I was hesitant doing that because the PR is for refactoring (+ some additional logging). This is kinda gray area. I will change according to your recommendation

Reply:
After thinking more I decided to rollback the change and address this comment in a separate PR by adding more metrics rather than changing the meaning of the current one in this PR. I discussed this with David offline and we agreed on the approach.